prow/gcsupload/run_test.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcsupload
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/diff"
+
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+func TestOptions_AssembleTargets(t *testing.T) {
+	var testCases = []struct {
+		name     string
+		options  Options
+		paths    []string
+		extra    map[string]gcs.UploadFunc
+		expected []string
+	}{
+		{
+			name: "no extra paths should upload infra files",
+			options: Options{
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			expected: []string{
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "extra paths should be uploaded under job dir",
+			options: Options{
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			extra: map[string]gcs.UploadFunc{
+				"something": gcs.DataUpload(strings.NewReader("data")),
+				"else":      gcs.DataUpload(strings.NewReader("data")),
+			},
+			expected: []string{
+				"pr-logs/pull/org_repo/1/job/build/something",
+				"pr-logs/pull/org_repo/1/job/build/else",
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "literal files should be uploaded under job dir",
+			options: Options{
+				Items: []string{"something", "else"},
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			paths: []string{"something", "else", "notforupload"},
+			expected: []string{
+				"pr-logs/pull/org_repo/1/job/build/something",
+				"pr-logs/pull/org_repo/1/job/build/else",
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "directories should be uploaded under job dir",
+			options: Options{
+				Items: []string{"something"},
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
```

Comment:
Can we test the other path strategies? Is it possible for this to be a default value?

Reply:
Because that's how integration tests are supposed to work?

IMO it is good to have `pod-utils/gcsupload` do the full cross-product validation, but it would be good if the integration test validates that we can modify the path strategy (meaning there is at least one test which uses a different strategy than all the others).